src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.Utils.cs

```diff
@@ -15,19 +19,20 @@ internal const
 
         public static int Compare(ReadOnlySpan<uint> left, ReadOnlySpan<uint> right)
         {
-            if (left.Length < right.Length)
-                return -1;
-            if (left.Length > right.Length)
-                return 1;
+            Debug.Assert(left.Length == 0 || left[^1] > 0);
+            Debug.Assert(right.Length == 0 || right[^1] > 0);
+
+            if (left.Length != right.Length)
+                return left.Length < right.Length ? -1 : 1;
+
+            ref uint rightPtr = ref MemoryMarshal.GetReference(right);
 
             for (int i = left.Length - 1; i >= 0; i--)
```

Comment:
Perhaps, it's better to wait (and ping) for https://github.com/dotnet/runtime/issues/9505 to fix bound checks in reversed loops instead of introducing yet another Unsafe magic we generally try to avoid now? Is this method perf critical?

Reply:
offline sync with @jakobbotsch: it should be handled by loop cloning, but loop cloning is currently never happening for Spans (it's included in 9.0 plan)

Reply:
I fixed Unsafe magic. The execution speed is the same as Unsafe magic.